### ISSUE TITLE
[Feature] Improve Kimi-K3 EAGLE3 acceptance across representative decode states

### ISSUE THREAD

**Issue by raikonenfnu** [lightseekorg/tokenspeed#1305]
## Checklist

- [x] I searched existing issues and open PRs before filing this request.

## Problem and motivation

Kimi-K3 EAGLE3 acceptance varies substantially across decode states. A raw random-completion benchmark can report high acceptance and decode throughput while producing a pathological two-token loop, masking much lower acceptance on normal chat output.

On TokenSpeed `ace99979b18a207e3bddb7bf643601693dc69f09`, using `lightseekorg/kimi-k3-eagle3-mla` with `topk=1`, 3 draft steps, and 4 draft tokens:

| Workload/state | Decoded tokens/iteration | Speculative accept rate | Decode tokens/s/user |
|---|---:|---:|---:|
| 50K random raw completion, 500 output tokens | 3.93 | 74.6% | 149.37 |
| 50K random chat completion, 500 output tokens | 1.94 | 48.5% | 74.74 |
| Representative AIME26 prompt, greedy, 1K output tokens | 2.99 | approximately 66% | not measured in this ablation |

The high raw-completion result is not representative: all 500 output tokens alternate between only two retokenized tokens (`�` and newline), with two unique bigrams. The chat-templated output is normal prose, with 282 unique retokenized tokens and 450 unique bigrams in the sampled response.

A controlled fixed-content ablation localized almost the entire change to the 14-token K3 assistant/thinking suffix:

| Fixed input content | Engine `Accept_num_tokens_avg` |
|---|---:|
| Raw prompt | 3.868217 |
| Raw prompt + assistant/thinking suffix | 2.079167 |
| Full K3 wrapper (74-token prefix + 14-token suffix) | 2.096639 |

Sending the identical rendered prompt through `/v1/chat/completions` and `/v1/completions` produced the same `Accept_num_tokens_avg=1.949219`, ruling out endpoint-specific execution.

Prefix caching also preserved acceptance exactly:

- Raw prompt: 3.868217 before and after reusing 49,920 of 50,000 prompt tokens.
- Full chat prompt: 2.096639 before and after reusing 50,048 of 50,088 prompt tokens.
- AIME26 prompt: 2.991018 before and after a 128-token prefix-cache hit.

This does not appear to be a target-correctness regression or a prefix-cache bug. It is an EAGLE3 quality/performance opportunity: improve draft agreement on representative chat, reasoning, agentic, and long-context states, and prevent degenerate synthetic output from hiding low-acceptance regimes.

## Reproduction

Server:

```bash
tokenspeed serve \
  --model moonshotai/Kimi-K3 \
  --tokenizer moonshotai/Kimi-K3 \
  --served-model-name kimi-k3 \
  --trust-remote-code \
  --reasoning-parser kimi_k3 \
  --tp 8 \
  --ep-size 8 \
  --init-expert-location trivial \
  --ep-dispatch-algorithm static \
  --moe-backend auto \
  --attention-backend gluon \
  --drafter-attention-backend gluon \
  --sampling-backend greedy \
  --kv-cache-dtype fp8 \
  --mm-encoder-tp-mode data \
  --max-model-len 65536 \
  --max-num-seqs 1 \
  --chunked-prefill-size 8192 \
  --max-prefill-tokens 8192 \
  --gpu-memory-utilization 0.92 \
  --cudagraph-capture-sizes 1 \
  --max-cudagraph-capture-size 1 \
  --disable-prefill-graph \
  --enable-prefix-caching \
  --disable-kvstore \
  --enable-cache-report \
  --speculative-algorithm EAGLE3 \
  --speculative-draft-model-path lightseekorg/kimi-k3-eagle3-mla \
  --speculative-num-steps 3 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 4 \
  --host 127.0.0.1 \
  --port 8000
```

Raw-completion control:

```bash
evalscope perf \
  --model kimi-k3 \
  --api openai \
  --url http://127.0.0.1:8000/v1/completions \
  --tokenizer-path moonshotai/Kimi-K3 \
  --dataset random \
  --min-prompt-length 50000 \
  --max-prompt-length 50000 \
  --min-tokens 500 \
  --max-tokens 500 \
  --dataset-offset 1110 \
  --seed 7 \
  --parallel 1 \
  --number 1 \
  --stream \
  --temperature 0 \
  --extra-args '{"ignore_eos":true}' \
  --no-apply-chat-template \
  --no-test-connection
```

Chat case:

```bash
evalscope perf \
  --model kimi-k3 \
  --api openai \
  --url http://127.0.0.1:8000/v1/chat/completions \
  --tokenizer-path moonshotai/Kimi-K3 \
  --dataset random \
  --min-prompt-length 50000 \
  --max-prompt-length 50000 \
  --min-tokens 500 \
  --max-tokens 500 \
  --dataset-offset 1110 \
  --seed 7 \
  --parallel 1 \
  --number 1 \
  --stream \
  --temperature 0 \
  --extra-args '{"ignore_eos":true}' \
  --apply-chat-template \
  --no-test-connection
```

The two EvalScope commands demonstrate the state-dependent acceptance. For strict endpoint isolation, render the chat template client-side and submit that identical rendered string to `/v1/completions`; it reproduces the chat endpoint's acceptance.

## Proposed solution

1. Add a representative Kimi-K3 EAGLE3 acceptance suite covering:
   - chat/reasoning and AIME-style prompts;
   - agentic/tool-use and code generation;
   - multilingual output;
   - long-context position buckets;
   - cold prompt, chunked prefill, prefix-cache hit, and decode graph/batch-size transitions.
2. Report acceptance per draft position (1 through 4), output-phase windows (first 64 tokens and steady state), and request-level distribution—not only a global average.
3. Compare draft and target logits/state at the first rejected token, checking the configured auxiliary captures `[2, 46, 90]`, control-token boundaries, RoPE/MLA position state, chunked-prefill handoff, and prefix-cache replay.
4. Improve the drafter/runtime integration or checkpoint quality on the representative suite without changing the target distribution or regressing AIME/agentic accuracy.
5. Consider an adaptive policy that reduces draft depth or disables speculation when rolling acceptance is below the measured break-even point.
6. Add a benchmark guard/report for extremely low output diversity so pathological loops cannot be presented as representative speculative-decoding wins.

Suggested initial success criteria:

- Demonstrate a statistically stable acceptance improvement on AIME/agentic/chat workloads across concurrency and context-length buckets.
- Preserve exact greedy target output where exact comparison is applicable, and preserve evaluation accuracy under sampling.
- Show end-to-end TPOT improvement, not only improved kernel or draft-model time.

## Environment

```text
Python: 3.10.12
ROCm available: True
GPU count: 8 (gfx950 / compute capability 9.5)
HIP: 7.14.60850
PyTorch: 2.13.0+rocm7.2
tokenspeed: 0.1.0
tokenspeed-kernel: 0.1.3.dev20260828+gitace99979
tokenspeed-kernel-amd: 0.1.3
tokenspeed-scheduler: 0.1.9
tokenspeed-triton: 3.8.10.post20260721
transformers: 5.12.0
```

**Comment by raikonenfnu** [comment]
Exact Kimi-K3 assistant/thinking generation suffix used in the controlled ablation:

```text
<|close|>message<|sep|><|end_of_msg|><|open|>message role="assistant"<|sep|><|open|>think<|sep|>
```

This is 14 tokens with the model tokenizer:

```text
[163588, 2778, 163589, 163586, 163587, 2778, 6244,
 878, 69702, 1, 163589, 163587, 39964, 163589]
```

| # | Token ID | Decoded token |
|---:|---:|---|
| 1 | 163588 | `<|close|>` |
| 2 | 2778 | `message` |
| 3 | 163589 | `<|sep|>` |
| 4 | 163586 | `<|end_of_msg|>` |
| 5 | 163587 | `<|open|>` |
| 6 | 2778 | `message` |
| 7 | 6244 | ` role` |
| 8 | 878 | `="` |
| 9 | 69702 | `assistant` |
| 10 | 1 | `"` |
| 11 | 163589 | `<|sep|>` |
| 12 | 163587 | `<|open|>` |
| 13 | 39964 | `think` |
| 14 | 163589 | `<|sep|>` |

Semantically, this suffix closes the user message, opens the assistant message, and enters the thinking channel. Appending just this suffix to the otherwise identical raw 50K prompt changed the engine's `Accept_num_tokens_avg` from `3.868217` to `2.079167`; the full chat wrapper measured `2.096639`.
